content/english/blog/balancing_nfr_coupling.md

```diff
@@ -0,0 +1,69 @@
+
+![Coupling Spectrum](/images/blogs/coupling_spectrum.png)
+
+As illustrated, there is a spectrum of coupling from utter and total tight coupling to free and unfettered, uncoupled design. On the far right, all of our components are stuck together, then crammed into a box and sealed. On the far left, we can end up with thousands of little tiny pieces that are so flexible that maintenance and discovery is a nightmare. Without me saying it, I bet you can imagine examples of current technologies that sit on either end of this spectrum.
+
+What we want, rather than extremes, is the _right_ balance. I want to write _just_ the code that satisfies my functional requirements, and then let other aspects of my system take care of satisfying the NFRs. To satisfy the above example, I want to write code that utilizes _some_ key-value store, _some_ SQL database, _some_ web server. I can then run that component against a test environment on my laptop, a medium-sized environment in the CI tests, and then massive scale when in production. More importantly, _while in production_, I can even change the deployment shape to suit changes in my user base, demand, load, etc.
+
+When we separate the functional and non-functional requirements across specific [contracts](https://en.wikipedia.org/wiki/Design_by_contract), then we can hopefully achieve the right balance between coupling. You might look at the word _contract_ here and think we're just going back to anti-corruption layers or abstraction libraries inside our code. The difference is subtle but vital: requiring that the thing that satisfies the contract _not be a part of your code_ is the magic sauce that balances the coupling scales.
+
+If we own our dependencies, and the implementation of a contract is _not_ a tightly coupled dependency, then we can finally say that, as developers, we don't own the non-functional requirements. NFRs are still mandatory, still important, and still have to be dealt with. However, if our _unit of deployment_ doesn't own them, then they are free to change, update, and move on their own cadence, _for their own reasons_.
+
+All of this lead-up and background hopefully explains why wasmCloud has chosen the abstractions we've chosen. We know that if we write a microservice that requires access to an **S3** bucket, it's going to be a huge pain to work with locally (even with tools like _minio_). The functional requirements of my code are "store this data in a blob store". Why should we be forced to decide on S3 and then live with that decision for the lifetime of the component? 
```

Comment:
Good opportunity for a heading here to separate the lead-up to the wasmCloud part